aptos-move/block-executor/src/lib.rs

```diff
@@ -147,3 +150,79 @@ mod txn_last_input_output;
 #[cfg(test)]
 mod unit_tests;
 pub mod view;
+
+/// In standard BlockSTM, the internal states assume contiguous transactions indices.
+/// To also support sharded execution (where each shard gets a subsequence of the transactions
+/// and possibly waits for transaction results from other shards),
+/// an index mapping is need as an additional input to our existing BlockSTM implementation.
+#[derive(Clone)]
+pub struct IndexMapping {
+    pub indices: Vec<TxnIndex>,
+    /// A TxnIndex -> local position mapping.
+    /// Currently implemented as a `Vec` of size equal to the block size, assuming it's not too large.
+    pub positions_by_index: Vec<usize>,
+    /// A fake index representing the end of the transaction list.
+    pub end_index: TxnIndex,
+}
+
+impl IndexMapping {
+    /// Create Positions by TxnIndex mapping from the TxnIndex list.
+    pub fn inverse(block_size: usize, indices: &Vec<TxnIndex>) -> Vec<usize> {
+        let mut ret = vec![usize::MAX; block_size];
+        for (pos, &index) in indices.iter().enumerate() {
+            ret[index as usize] = pos;
+        }
+        ret
+    }
+
+    pub fn new(indices: Vec<TxnIndex>, block_size: usize) -> Self {
+        let positions_by_index = Self::inverse(block_size, &indices);
+        Self {
+            indices,
+            positions_by_index,
+            end_index: TxnIndex::MAX,
+        }
+    }
+
+    pub fn wrap(indices: Vec<TxnIndex>, positions_by_idx: Vec<usize>) -> Self {
+        let end_index = TxnIndex::MAX;
+        Self {
+            indices,
+            positions_by_index: positions_by_idx,
+            end_index,
+        }
+    }
+
+    pub fn new_unsharded(block_size: usize) -> Self {
+        Self {
+            indices: (0..(block_size as TxnIndex)).collect(),
+            positions_by_index: (0..block_size).collect(),
+            end_index: block_size as TxnIndex,
+        }
+    }
+
+    pub fn next_index(&self, index: TxnIndex) -> TxnIndex {
```

Comment:
nit: this function can be a bit simplified removing unnecessary multiple level of if-else nesting and some refactoring to make it better readable.

```

   pub fn is_end_marker(index: TxnIndex) -> bool {
      index == END_INDEX_MARKER
   } 
    
   pub fn is_last_index(&self, index: TxnIndex) -> bool {
        self.positions_by_index[index as usize] == self.indices.len() - 1 
    }

    pub fn next_index(&self, index: TxnIndex) -> TxnIndex {
        if is_end_marker(index) || is_last_index(index) {
            return END_INDEX_MARKER
        }
        let pos = self.positions_by_index[index as usize];
        return self.indices[pos + 1]    
    }
  ```

Reply:
fixed